File: typescript/packages/core/src/process/stdio.ts
import { Channel, JobConsole } from '../shell/console/index.ts'
import { PipeConsole } from '../shell/console/pipe.ts'
import { PipeClosed } from '../shell/errors.ts'

export class ProcessInput {
  private readonly pipe = new PipeConsole()
  private closed = false
  bytesRead = 0

  async write(data: Uint8Array): Promise<void> {
    if (data.byteLength === 0) return
    if (this.closed) throw new PipeClosed()
    for (let start = 0; start < data.byteLength; start += 65536)
      await this.pipe.emit(Channel.STDOUT, data.slice(start, start + 65536))
  }

  close(): void {
    this.closed = true
    this.pipe.end()
  }
  stop(): void {
    this.close()
    this.pipe.closeReader()
  }
  async *stream(): AsyncIterable<Uint8Array> {
    for await (const chunk of this.pipe.stream()) {
      this.bytesRead += chunk.byteLength
      yield chunk
    }
  }
}

export class ProcessOutput extends JobConsole {
  readonly stdout = new ProcessInput()
  readonly stderr = new ProcessInput()

  constructor(private readonly mergeStderr = false) {
    super()
  }

  override async emit(channel: Channel, data: Uint8Array): Promise<void> {
    await (channel === Channel.STDERR && !this.mergeStderr ? this.stderr : this.stdout).write(data)
  }

  end(): void {
    this.stdout.close()
    this.stderr.close()
  }
  stop(): void {
    this.stdout.stop()
    this.stderr.stop()
  }
}

File: python/mirage/process/stdio.py
from collections.abc import AsyncIterator

from mirage.shell.console import Channel, JobConsole
from mirage.shell.console.pipe import PipeConsole
from mirage.shell.errors import PipeClosed


class ProcessInput:

    def __init__(self) -> None:
        self.pipe = PipeConsole()
        self.closed = False

    async def write(self, data: bytes) -> None:
        if not data:
            return
        if self.closed:
            raise PipeClosed()
        for start in range(0, len(data), 65536):
            await self.pipe.emit(Channel.STDOUT, data[start:start + 65536])

    def close(self) -> None:
        self.closed = True
        self.pipe.end()

    def stop(self) -> None:
        self.close()
        self.pipe.close_reader()

    def stream(self) -> AsyncIterator[bytes]:
        return self.pipe.stream()


class ProcessOutput(JobConsole):

    def __init__(self) -> None:
        super().__init__()
        self.stdout = ProcessInput()
        self.stderr = ProcessInput()

    async def emit(self, channel: Channel, data: bytes) -> None:
        target = self.stderr if channel == Channel.STDERR else self.stdout
        await target.write(data)

    def end(self) -> None:
        self.stdout.close()
        self.stderr.close()

    def stop(self) -> None:
        self.stdout.stop()
        self.stderr.stop()
